service-utils/gen-ssl.sh

```diff
@@ -1,49 +1,129 @@
-path_to_etc=/etc
-openssl_conf_path=/etc/ssl/ct-openssl.conf
+# Die on errors immediately.
+set -e
 
-echo "Setting up credential directory structure..."
-mkdir -p $path_to_etc/creds/root/certs
-mkdir -p $path_to_etc/creds/root/keys
-mkdir -p $path_to_etc/creds/services/certs
-mkdir -p $path_to_etc/creds/services/keys
+# First argument sets the location of /etc directory.
+path_to_etc=${1:?"An absolute path to /etc must be the first parameter."}
 
-mkdir tmp && cd tmp
+shift  # Cut off $1 so that the rest of $@ is a list of names.
+if [ -z "$*" ]; then
+    echo "No service names passed!"
+    exit 1
+fi
 
-echo "Generating root certs..."
-openssl genrsa -passout pass:1111 -des3 -out ca.key.pem 4096
+# Allow environment to override openssl config file.
+openssl_conf_file=${CT_OPENSSL_CONF_FILE:-/etc/ssl/ct-openssl.conf}
+if [ ! -f  "${openssl_conf_file}" ]; then
+    echo "No openssl config file at '${openssl_conf_file}'."
+    exit 1
+fi
 
-openssl req -passin pass:1111 -new -x509 -days 7300 -key ca.key.pem -out ca.cert.pem -subj  "/C=FR/ST=Paris/L=Paris/O=Test/OU=Test/CN=root"
+ROOT_CERTS=${path_to_etc}/creds/root/certs
+ROOT_KEYS=${path_to_etc}/creds/root/keys
+SVC_CERTS=${path_to_etc}/creds/services/certs
+SVC_KEYS=${path_to_etc}/creds/services/keys
 
-echo "Generating intermediate certs..."
-openssl genrsa -passout pass:1111 -des3 -out ca-intermediate.key.pem 4096
+TMP="tmp-gen-ssl-$$"  # Temporary work dir name.
```

Comment:
The `$$` makes each script invocation make a separate dir.If an invocation fails for whatever reason, and the temporary dir is not removed, following invocations still get a clean new dir.